### PR TITLE
Codegen address prefixes

### PR DESCRIPTION
**Pull request by hewigovens**
Changes:

- Generate p2pkh / p2sh / hrp from coin definition (`coins.json`)
- Fix Decred p2sh lock script
- Fix https://github.com/TrustWallet/wallet-core/issues/505

### CODE REVIEW

**Comment by hewigovens** [comment]
@alejandro-isaza would you mind to take a look? Thanks

File: codegen/lib/templates/coins.cpp.erb [hunk 1]
```diff
@@ -83,6 +83,60 @@ enum TWPublicKeyType TW::publicKeyType(TWCoinType coin) {
...
+        .each do |prefix|
+-%>
+        <%= prefix %>,
+<% end -%>
+    });
+}
+
+enum TWHRP TW::hrp(TWCoinType coin) {
+    switch (coin) {
+<%  coins.filter{ |coin| coin.has_key?('hrp') }.each do |coin| -%>
```
Comment (vikmeup) [inline codegen/lib/templates/coins.cpp.erb]: Any chance we can add test for it? in the case somehow break it

File: codegen/lib/templates/coins.cpp.erb [hunk 2]
```diff
@@ -83,6 +83,60 @@ enum TWPublicKeyType TW::publicKeyType(TWCoinType coin) {
...
+        return <%= coin['p2shPrefix'] %>;
+<% end -%>
+    default: return 0;
+    }
+}
+
+std::set<byte> TW::p2shPrefixes() {
+    return std::set<byte>({
+<%  coins.filter{ |coin| coin.has_key?('p2shPrefix') && coin['id'] != 'decred' }
+        .map{ |coin| coin['p2shPrefix'] }
```
Comment (vikmeup) [inline codegen/lib/templates/coins.cpp.erb]: what is so special about decred?

File: codegen/lib/templates/coins.cpp.erb (hunk 1 above)
Comment (vikmeup) [reply codegen/lib/templates/coins.cpp.erb]: to each function

File: codegen/lib/templates/coins.cpp.erb (hunk 2 above)
Comment (hewigovens) [reply codegen/lib/templates/coins.cpp.erb]: Decred has two bytes prefix, we need to find another way to handle it

File: codegen/lib/templates/coins.cpp.erb (hunk 1 above)
Comment (hewigovens) [reply codegen/lib/templates/coins.cpp.erb]: Yes, we have `TWHRPTests`, just make sure test new added hrp

**Pull request merged**